### ISSUE TITLE
Proposal to change the offsets for the EmbeddingBag operator in PyTorch

### ISSUE THREAD

**Issue by jianyuh** [pytorch/pytorch#29019]
## 🚀 Feature
The PyTorch EmbeddingBag operator computes sums or means of “bags” of embeddings, without instantiating the intermediate embeddings. Assuming the number of “bags” is B. Let’s take a look at a simple PyTorch EmbeddingBag operator example:

```
# an Embedding module containing 10 tensors of size 3
embedding_sum = nn.EmbeddingBag(10, 3, mode='sum')
# a batch of 2 samples of 4 indices each, so the number of bags B = 2 in this example
input = torch.LongTensor([1,2,4,5,4,3,2,9])
offsets = torch.LongTensor([0,4])
output = embedding_sum(input, offsets)
print(output)

tensor([[-0.8861, -5.4350, -0.0523],
[ 1.1306, -2.5798, -1.0044]])
```

Here is the simplest version of PyTorch EmbeddingBag API:

```
embedding_sum = torch.nn.EmbeddingBag(num_embeddings, embedding_dim, mode='sum')
```

* num_embedding: the size of the dictionary of embeddings.
* embedding_dim: the size of each embedding vector.
* mode=‘sum’: computes the (weighted) sum.

Here is the usage for this EmbeddingBag operator:
```
output = embedding_sum(input, offsets)
```

`input` is 1D of shape (N), it will be treated as a concatenation of multiple bags (sequences).
 `offsets` is required to be a 1D tensor containing the starting index positions of each bag in `input`. Therefore, for `offsets` of shape (B), input will be viewed as having B bags. Empty bags (i.e., having 0-length) will have returned vectors filled by zeros.

Currently offsets has the size the same as the number of bags B. It can be computed as the prefix sum of the individual length for each bag.

* offsets[0] = 0
* offsets[i] = offsets[i-1] + lengths[i-1] for 1 <= i <= B.

We need to compute the length for each individual bag for the implementation of EmbeddingBag under the hood. In order to compute the length for each individual bag, we need to calculate as the following:

* lengths[i] = offsets[i+1] - offsets[i], for 0 <= i <= B - 2.
* lengths[B] = len(input) - offsets[B]

## Motivation

we would like to avoid the special handler for the last element. This make the usage of `aten::parallel_for` API easier if we want to have the intra-op parallelism for EmbeddingBag operator. This is also consistent with the normal CSR/CSC format.

## Pitch

*Can we add one more element in the offsets?* The last element is equal to the size of the input.

*Benefits:*
We can compute the length from the offsets with the following

* lengths[i] = offsets[i+1] - offsets[i], for 0 <= i <= B-1.

In this way, we avoid the special handler for the last element. This make the usage of aten::parallel_for API easier if we want to have the intra-op parallelism for EmbeddingBag operator. This is also consistent with the normal CSR/CSC format.

*Options:*
We can add the offsets in the top level API (i.e., PyTorch EmbeddingBag operator), or just theEmbeddingLookUp APIs

* Top-level API: We need to change the PyTorch API, which might break things (backward compatibility issue).
* EmbeddingLookUp perfkernel API: we have to do the offsets (size of B) → offsets (size of B+1) conversion, which might involve some overhead.
* Add a new frontend interface to avoid backward compatibility issue.

*Question*: Can we just check if the last element is equivalent to the size of "input" (which is the embedding id list) to distinguish whether offsets have (B+1) element or (B) elements? Specifically, when we only have (B) elements, we attach one element in the end before calling the new EmbeddingLookUp API (which takes-(B+1) element offsets).

*Answer* (From @jspark1105 ): a challenge is the number of elements in offset is the only way we know the batch size. We can't also rely on offset[-1] == lengths(indices) because it's possible we can have 0 length at the end.
Consider an example of input = [1,2,4,5,4,3,2,9] and offset = [0,4,8] used for old interface (the last sample in the batch has length 0). In the new interface, we'd have offset = [0,4,8,8] so we can't tell whether if the new interface is used by just looking at the last element of offset.
We'd need a new frontend interface to avoid backward compatibility issues.

cc @dzhulgakov @jspark1105 @yinghai @bddppq @JianpingChen066 

Related PRs blocked by this issue:
https://github.com/pytorch/pytorch/pull/27477
https://github.com/pytorch/pytorch/pull/24385
https://github.com/pytorch/pytorch/pull/24945
https://github.com/pytorch/pytorch/pull/24944

**Comment by colesbury** [comment]
This would be a significant breaking change, right?

**Comment by colesbury** [comment]
Assuming the answer to the above is "yes", a follow up question: Is there a way to implement this so that both offset formats can be supported in a release and the existing offsets can be deprecated over one or more releases?

**Comment by jianyuh** [comment]
> This would be a significant breaking change, right?

Yes: If we just change the same EmbeddingBag API and add one more dimension to the `offsets`, it will not provide the backward compatibility.

**Comment by jianyuh** [comment]
> Assuming the answer to the above is "yes", a follow up question: Is there a way to implement this so that both offset formats can be supported in a release and the existing offsets can be deprecated over one or more releases?

See the question and answer in the end of the issue: Consider an example of input = [1,2,4,5,4,3,2,9] and offset = [0,4,8] used for old interface (the last sample in the batch has length 0). In the new interface, we'd have offset = [0,4,8,8] so we can't tell whether if the new interface is used by just looking at the last element of offset.
It appears that we need a new frontend interface to avoid the backward compatibility issue.

**Comment by yinghai** [comment]
Yeah, we need a new flag in the frontend to have this new change. And this new change is not just for convenience. To lower this op to accelerator, we usually create a larger buffer and zero pad. For example, we want to look up indices [0, 1, 2, 3], but since we allocate the indices as max size of 8, the input to accelerator becomes [0, 1, 2, 3, 0, 0, 0, 0]. Since `0` is a valid lookup index, we don't know where we are going to end just from current last offset. One argument is that we can pad something like `-1` to have [0, 1, 2, 3, -1, -1, -1, -1]. But padding -1 is probably not as efficient and we have to add more branching logic to check whether we hit -1 or not.

**Comment by yinghai** [comment]
For ui, we might want to have something like
```
embedding_sum = nn.EmbeddingBag(10, 3, mode='sum', new_offset=True)
```
And stub a default as `False`.

**Comment by jspark1105** [comment]
> Yeah, we need a new flag in the frontend to have this new change. And this new change is not just for convenience. To lower this op to accelerator, we usually create a larger buffer and zero pad. For example, we want to look up indices [0, 1, 2, 3], but since we allocate the indices as max size of 8, the input to accelerator becomes [0, 1, 2, 3, 0, 0, 0, 0]. Since `0` is a valid lookup index, we don't know where we are going to end just from current last offset. One argument is that we can pad something like `-1` to have [0, 1, 2, 3, -1, -1, -1, -1]. But padding -1 is probably not as efficient and we have to add more branching logic to check whether we hit -1 or not.

@yinghai this discussion is more about having one more element in offset not about indices. Maybe we can separate discussion about paddings for accelerators from this?

**Comment by jspark1105** [comment]
> Assuming the answer to the above is "yes", a follow up question: Is there a way to implement this so that both offset formats can be supported in a release and the existing offsets can be deprecated over one or more releases?

I think @yinghai's suggestion on adding new_offset=True can be a way to do what you suggested?

**Comment by colesbury** [comment]
> I think yinghai's suggestion on adding new_offset=True can be a way to do what you suggested?

Yes, sounds good

**Comment by yinghai** [comment]
> @yinghai this thread is more about having one more element in offset not about indices. Maybe we can separate discussion about paddings for accelerators from this?

Basically, if we go with proposed method, it will solve the padding on indices issue too.

**Comment by dzhulgakov** [comment]
> Basically, if we go with proposed method, it will solve the padding on indices issue too.
There's still a corner case `0,0,0,0` - is it padded sequence of empty lists or just an empty list?

Since we're discussing changing embedding_bag api - we can also consider `length` encoding instead of offsets. Pro: it has the same size without extra element, support concat to concatenate two lists. Cons: I think majority of kernels actually need to compute prefixsum anyway (and the proposed "new_prefix" actually matches CSR format directly: https://en.wikipedia.org/wiki/Sparse_matrix#Compressed_sparse_row_(CSR,_CRS_or_Yale_format) )

**Comment by yinghai** [comment]
For offsets [0, 0, 0, 0], we would expect output from acclerator to be all zeros in either case? If we are going to have a flag to change the flavor of embedding bag, maybe having lengths isn't a bad idea, except for that you need to do prefix sum basically defeats the purpose of the parallel for that the proposal is pitching.

**Comment by jackm321** [comment]
Between the proposed method, padding with -1 and using lengths, the proposed change seems like the best one because it solves both the accelerator and parallel_for use cases at once, doesn't require prefix sum compute everywhere, and has only the drawbacks of sizeof(offsets) == 1 + sizeof(outputs) and it introduces breaking changes which can be mitigated with a flag.

**Comment by jianyuh** [comment]
Request for comments in #27477.

**Comment by jianyuh** [comment]
Closing as https://github.com/pytorch/pytorch/pull/27477 has landed.

**Comment by jianyuh** [comment]
Also link https://github.com/pytorch/pytorch/pull/42215 for mean and max support.
